snapcraft/linters/classic_linter.py

```diff
@@ -0,0 +1,136 @@
+        installed_base_path = Path(f"/snap/{self._snap_metadata.base}/current")
+
+        # ELF files must be patched if confinement is classic or libc is staged
+        try:
+            # If libc is staged we'll find a dynamic linker in the current path
+            # (which contains the unpackaged primed payload). In this case, at
+            # runtime the linker will be in the installed snap path.
+            linker = elf_utils.get_dynamic_linker(
+                root_path=current_path, snap_path=installed_snap_path
+            )
+            is_libc_staged = True
+            issue = LinterIssue(
+                name=self._name,
+                result=LinterResult.OK,
+                text="Snap contains staged libc.",
+            )
+        except errors.DynamicLinkerNotFound:
+            # Otherwise look for the host linker, which should match the base
+            # system linker. At runtime use the linker from the installed base
+            # snap.
+            linker = elf_utils.get_dynamic_linker(
+                root_path=Path("/"), snap_path=installed_base_path
+            )
+            is_libc_staged = False
+            issue = LinterIssue(
+                name=self._name,
+                result=LinterResult.OK,
+                text="Snap confinement is set to classic.",
+            )
+
+        # No issues to report if confinement is not classic and libc is not staged.
+        if not is_libc_staged and self._snap_metadata.confinement != "classic":
+            return []
+
+        issues = [issue]
+        elf_files = elf_utils.get_elf_files(current_path)
+        patcher = Patcher(dynamic_linker=linker, root_path=current_path.absolute())
+        soname_cache = SonameCache()
+
+        for elf_file in elf_files:
+            # Skip linting files listed in the ignore list.
+            for pattern in self._lint.ignore.files:
+                if fnmatch.fnmatch(str(elf_file.path), pattern):
+                    emit.debug(
+                        f"ClassicLinter: skip file {str(elf_file.path)!r} "
+                        f"(matches {pattern!r})"
+                    )
+
+            arch_triplet = elf_utils.get_arch_triplet()
+
+            elf_file.load_dependencies(
+                root_path=current_path,
+                base_path=installed_base_path,
+                content_dirs=set(),  # FIXME: obtain content dirs
+                arch_triplet=arch_triplet,
+                soname_cache=soname_cache,
+            )
+
+            self._check_elf_interpreter(elf_file, linker=linker, issues=issues)
+            self._check_elf_rpath(elf_file, patcher=patcher, issues=issues)
+
+        return issues
+
+    def _check_elf_interpreter(
+        self, elf_file: ElfFile, *, linker: str, issues: List[LinterIssue]
+    ) -> None:
+        """Check ELF executable interpreter is set to base or snap linker."""
+        if elf_file.interp != linker:
+            issue = LinterIssue(
```

Comment:
would be good to have a url with more context, I created a separate issue for that